File: app/src/main/java/com/minerdev/greformanager/HttpConnection.java
package com.minerdev.greformanager;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.util.Log;
import android.widget.Toast;

import com.android.volley.AuthFailureError;
import com.android.volley.NetworkResponse;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.google.gson.Gson;
import com.google.gson.JsonObject;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class HttpConnection {
    private HttpConnection() {

    }

    public static HttpConnection getInstance() {
        return Holder.INSTANCE;
    }

    public void receive(Context context, String uri, OnReceiveListener listener) {
        String serverUri = Constants.getInstance().DNS + "/" + uri;
        Log.d("SEND_DATA", serverUri);
        makeRequest(context, Request.Method.GET, serverUri, null, listener);
    }

    public void send(Context context, int method, String uri, JsonObject data, OnReceiveListener listener) {
        String serverUri = Constants.getInstance().DNS + "/" + uri;
        String json = data.toString();
        Log.d("SEND_DATA", json);
        makeRequest(context, method, serverUri, json, listener);
    }

    public void send(Context context, int method, String uri, House.ParcelableData data, OnReceiveListener listener) {
        String serverUri = Constants.getInstance().DNS + "/" + uri;
        Gson gson = new Gson();
        String json = gson.toJson(data);
        Log.d("SEND_DATA", json);
        makeRequest(context, method, serverUri, json, listener);
    }

    public void send(Context context, int method, String uri, ArrayList<Uri> imageUris, OnReceiveListener listener) {
        String serverUri = Constants.getInstance().DNS + "/" + uri;
        int position = 0;
        for (Uri imageUri : imageUris) {
            Log.d("SEND_DATA", imageUri.getPath());
            makeImageRequest(context, method, serverUri, imageUri, position, listener);
            position++;
        }
    }

    private void makeRequest(Context context, int method, String uri, String json, OnReceiveListener listener) {
        StringRequest request = new StringRequest(method, uri,
                response -> {
                    Toast.makeText(context, "데이터 전송 성공.", Toast.LENGTH_SHORT).show();
                    if (listener != null) {
                        listener.onReceive(response);
                    }
                },
                error -> Toast.makeText(context, error.getMessage(), Toast.LENGTH_SHORT).show()
        ) {
            @Override
            public byte[] getBody() throws AuthFailureError {
                if (json == null || json.equals(""))
                    return super.getBody();

                return json.getBytes(StandardCharsets.UTF_8);
            }

            @Override
            public String getBodyContentType() {
                if (json == null || json.equals(""))
                    return super.getBodyContentType();

                return "application/json";
            }
        };

        // 이전 결과가 있더라도 새로 요청
        request.setShouldCache(false);
        AppHelper.requestQueue.add(request);
    }


    private void makeImageRequest(Context context, int method, String uri, Uri imageUri, int position, OnReceiveListener listener) {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        File file = new File(getPathFromUri(context, imageUri));
        FileInputStream inputStream;
        byte[] buf = new byte[1024];
        int size;

        if (file.exists() && file.canRead()) {
            try {
                inputStream = new FileInputStream(file);
                while ((size = inputStream.read(buf)) != -1) {
                    byteArrayOutputStream.write(buf, 0, size);
                }

                inputStream.close();

            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        String title = System.currentTimeMillis() + ".jpg";
        VolleyMultipartRequest request = new VolleyMultipartRequest(method, uri,
                new Response.Listener<NetworkResponse>() {
                    @Override
                    public void onResponse(NetworkResponse response) {
                        try {
                            JSONObject obj = new JSONObject(new String(response.data));
                            Toast.makeText(context, obj.getString("message"), Toast.LENGTH_SHORT).show();

                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        Toast.makeText(context, error.getMessage(), Toast.LENGTH_LONG).show();
                        Log.e("GotError", "" + error.getMessage());
                    }
                }
        ) {
            @Override
            protected Map<String, String> getParams() throws AuthFailureError {
                Map<String, String> params = new HashMap<>();
                params.put("title", title);
                params.put("position", String.valueOf(position));
                return params;
            }

            @Override
            protected Map<String, DataPart> getByteData() {
                Map<String, DataPart> params = new HashMap<>();
                params.put("image", new DataPart(title, byteArrayOutputStream.toByteArray()));
                return params;
            }
        };

        // 이전 결과가 있더라도 새로 요청
        request.setShouldCache(false);
        AppHelper.requestQueue.add(request);
    }

    public String getPathFromUri(Context context, Uri uri) {
        Cursor cursor = context.getContentResolver().query(uri, null, null, null, null);
        cursor.moveToNext();
        String path = cursor.getString(cursor.getColumnIndex("_data"));
        cursor.close();

        return path;
    }

    public interface OnReceiveListener {
        void onReceive(String receivedData);
    }

    private static class Holder {
        public static final HttpConnection INSTANCE = new HttpConnection();
    }
}

File: app/src/main/java/com/minerdev/greformanager/HttpConnection.kt
package com.minerdev.greformanager

import android.net.Uri
import android.util.Log
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.StringRequest
import com.minerdev.greformanager.model.House
import com.minerdev.greformanager.model.Image
import com.minerdev.greformanager.utils.AppHelper
import com.minerdev.greformanager.utils.Constants
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.json.JSONObject
import java.nio.charset.StandardCharsets

class HttpConnection private constructor() {
    fun receive(uri: String, listener: OnReceiveListener?) {
        val serverUri: String = Constants.instance.DNS + "/api/" + uri
        Log.d("HTTP_DATA", "receive : $serverUri")
        makeJsonRequest(Request.Method.GET, serverUri, null, listener)
    }

    fun send(method: Int, uri: String, listener: OnReceiveListener?) {
        val serverUri: String = Constants.instance.DNS + "/api/" + uri
        Log.d("HTTP_DATA", "send : $serverUri")
        makeJsonRequest(method, serverUri, null, listener)
    }

    fun send(method: Int, uri: String, data: JSONObject, listener: OnReceiveListener?) {
        val serverUri: String = Constants.instance.DNS + "/api/" + uri
        val json = data.toString()
        Log.d("HTTP_DATA", "send : $json")
        makeJsonRequest(method, serverUri, json, listener)
    }

    fun send(method: Int, uri: String, data: House, listener: OnReceiveListener?) {
        val serverUri: String = Constants.instance.DNS + "/api/" + uri
        val format = Json { encodeDefaults = true }
        val json = format.encodeToString(data)
        val prettyJson = Json { prettyPrint = true; encodeDefaults = true; }
        Log.d("HTTP_DATA", "send : \n" + prettyJson.encodeToString(data))
        makeJsonRequest(method, serverUri, json, listener)
    }

    fun send(method: Int, uri: String, data: Image, listener: OnReceiveListener?) {
        val serverUri: String = Constants.instance.DNS + "/api/" + uri
        val format = Json { encodeDefaults = true }
        val json = format.encodeToString(data)
        val prettyJson = Json { prettyPrint = true; encodeDefaults = true; }
        Log.d("HTTP_DATA", "send : \n" + prettyJson.encodeToString(data))
        makeJsonRequest(method, serverUri, json, listener)
    }

    fun sendImage(method: Int, uri: String, data: Image, listener: OnReceiveListener?) {
        val serverUri: String = Constants.instance.DNS + "/api/" + uri
        val prettyJson = Json { prettyPrint = true }
        Log.d("HTTP_DATA", "send : \n" + prettyJson.encodeToString(data))
        makeImageRequest(method, serverUri, data, listener)
    }

    private fun makeJsonRequest(method: Int, uri: String, json: String?, listener: OnReceiveListener?) {
        val request: StringRequest = object : StringRequest(method, uri,
                Response.Listener { response ->
                    Log.d("HTTP_DATA", "makeHouseRequest response : $response")

                    if (!response.isNullOrEmpty()) {
                        listener?.onReceive(response)
                    }
                },
                Response.ErrorListener { error ->
                    Log.e("HTTP_DATA", "makeHouseRequest error : " + error.message)
                }
        ) {
            override fun getBody(): ByteArray {
                return if (json.isNullOrEmpty()) super.getBody() else json.toByteArray(StandardCharsets.UTF_8)
            }

            override fun getBodyContentType(): String {
                return if (json.isNullOrEmpty()) super.getBodyContentType() else "application/json"
            }
        }

        // 이전 결과가 있더라도 새로 요청
        request.setShouldCache(false)
        AppHelper.instance.requestQueue?.add(request)
    }

    private fun makeImageRequest(method: Int, uri: String, image: Image, listener: OnReceiveListener?) {
        val request: VolleyMultipartRequest = object : VolleyMultipartRequest(method, uri,
                Response.Listener { response ->
                    Log.d("HTTP_DATA", "makeImageRequest response")

                    response?.let {
                        if (it.data != null && it.data.isNotEmpty()) {
                            listener?.onReceive(String(it.data))
                        }
                    }
                },
                Response.ErrorListener { error ->
                    Log.e("HTTP_DATA", "makeImageRequest error : " + error.message)
                }
        ) {
            override fun getParams(): MutableMap<String, String> {
                val params: MutableMap<String, String> = HashMap()
                params["position"] = image.position.toString()
                params["thumbnail"] = image.thumbnail.toString()
                return params
            }

            override val byteData: Map<String, DataPart>
                get() {
                    val params: MutableMap<String, DataPart> = HashMap()
                    image.localUri?.let {
                        val imageUri = Uri.parse(image.localUri)
                        params["image"] = DataPart("image", AppHelper.instance.getByteArrayFromUri(imageUri))
                    }
                    return params
                }
        }

        // 이전 결과가 있더라도 새로 요청
        request.setShouldCache(false)
        AppHelper.instance.requestQueue?.add(request)
    }

    private fun makeImageRequest2(method: Int, uri: String, image: Image, listener: OnReceiveListener?) {
        val request = Request()
        val request: VolleyMultipartRequest = object : VolleyMultipartRequest(method, uri,
                Response.Listener { response ->
                    Log.d("HTTP_DATA", "makeImageRequest response")

                    response?.let {
                        if (it.data != null && it.data.isNotEmpty()) {
                            listener?.onReceive(String(it.data))
                        }
                    }
                },
                Response.ErrorListener { error ->
                    Log.e("HTTP_DATA", "makeImageRequest error : " + error.message)
                }
        ) {
            override fun getParams(): MutableMap<String, String> {
                val params: MutableMap<String, String> = HashMap()
                params["position"] = image.position.toString()
                params["thumbnail"] = image.thumbnail.toString()
                return params
            }

            override val byteData: Map<String, DataPart>
                get() {
                    val params: MutableMap<String, DataPart> = HashMap()
                    image.localUri?.let {
                        val imageUri = Uri.parse(image.localUri)
                        params["image"] = DataPart("image", AppHelper.instance.getByteArrayFromUri(imageUri))
                    }
                    return params
                }
        }

        // 이전 결과가 있더라도 새로 요청
        request.setShouldCache(false)
        AppHelper.instance.requestQueue?.add(request)
    }

    interface OnReceiveListener {
        fun onReceive(receivedData: String)
    }

    private object Holder {
        val INSTANCE = HttpConnection()
    }

    companion object {
        val instance: HttpConnection
            get() = Holder.INSTANCE
    }
}